### DISCUSSION TITLE
Is it possible to run this on raspberry-pi?

### DISCUSSION THREAD

**Discussion by gkns** [euzu/tuliprox#24] [Q&A] [answered]
My intention is to have this running this on my local Pi and use that as a proxy for the TV to filter content from an XSTREAM code enabled IPTV provider.

I checked running the released binary for Linux. However, there is an 'Exec format error'.
In the compile instructions there is nothing specific for Raspberry Pi. Any ideas?

I don't mind not having the UI.
I just need a headless service for filtering the playlist.

**Comment by euzu** [comment]
hi @gkns, i dont own a pi, thats why i cant test it, but it is possible to crosscompile  on your linux or windows machine the rust code for  arm cpu's.  The UI is running in a browser, if you get m3u-filter running, the ui will be available too.

I have searched and found 
[Cross Compiling Rust for the Raspberry Pi](https://chacin.dev/blog/cross-compiling-rust-for-the-raspberry-pi/)

Can you report if this works ?

**Comment by euzu** [comment]
I have crosscompiled a pi 2/3/4  binary:
[m3u-filter.zip](https://github.com/euzu/m3u-filter/files/13377033/m3u-filter.zip)

**Reply by gkns** [reply]
Thanks! I tried the attached executable, but it doesn't run.
Hit this issue: https://unix.stackexchange.com/questions/101429/no-such-file-or-directory-when-executing-a-cross-compiled-program-on-a-raspber

I have zero knowledge of rust/cargo.

I tried cross-compiling on Apple Silicon Mac and an Ubuntu machine.
On Apple Silicon, eventough I used a correct linker (I believe) - https://developer.arm.com/-/media/Files/downloads/gnu/13.2.rel1/binrel/arm-gnu-toolchain-13.2.rel1-darwin-arm64-arm-none-eabi.tar.xz?rev=73e10891de3d41e29e95ac2878742b74&hash=2FBE6B355E5A06A7A7A2BB51719BF0C2FAC4A138

The compilation itself didn't go through.

Hence I tried Ubuntu VM with : https://developer.arm.com/-/media/Files/downloads/gnu-a/10.3-2021.07/binrel/gcc-arm-10.3-2021.07-aarch64-arm-none-linux-gnueabihf.tar.xz?rev=eefffd05fe254e8190f4aafd17dbaee5&hash=5AC740AB267BAF170EA5A6F07B0ECECEC8B20F7D 

That failed for 329th package - m3u-filter itself on linking phase with errors like:
```
          /usr/bin/ld: /home/gopikrishnas/git-repos/m3u-filter/target/armv7-unknown-linux-gnueabihf/release/deps/m3u_filter-a3459ee102a68daa.m3u_filter.61b6e54f5fa0c6d9-cgu.0.rcgu.o: relocations in generic ELF (EM: 40)
          /usr/bin/ld: /home/gopikrishnas/git-repos/m3u-filter/target/armv7-unknown-linux-gnueabihf/release/deps/m3u_filter-a3459ee102a68daa.m3u_filter.61b6e54f5fa0c6d9-cgu.0.rcgu.o: error adding symbols: file in wrong format
          collect2: error: ld returned 1 exit status


error: could not compile `m3u-filter` (bin "m3u-filter") due to previous error
```
So overall still no-go.

**Reply by euzu** [reply]
Which raspberry Pi do you have ? I have compiled for 2/3/4

**Reply by gkns** [reply]
I have Pi4 model B

**Reply by gkns** [reply]
From: https://unix.stackexchange.com/questions/101429/no-such-file-or-directory-when-executing-a-cross-compiled-program-on-a-raspber

`If you are using crosstool-NG, have a look at the .config for the value of CT_ARCH_ARCH. For the raspberry pi, it should be "armv6j"1 -- or at least, that's what's working for me. There are other specifics, but I think that should be enough. Unfortunately, if it's wrong, you now have to rebuild.`

Not sure how you compiled it, But is the above worth considering?

**Reply by euzu** [reply]
i tried to compile a static binary but failed because of some missing libraries, and i couldn't find a solution.

**Comment by euzu** [comment]
Here is a cross comipled static binary  compile with `cross`
[m3u-filter_armv7_musleabihf.zip](https://github.com/euzu/m3u-filter/files/13379165/m3u-filter_armv7_musleabihf.zip)

Can you report if this works ?

**Reply by gkns** [reply]
Thank you! This executable is loading:
```
./m3u-filter --help
Extended M3U playlist filter

Usage: m3u-filter [OPTIONS]

Options:
  -c, --config <CONFIG>        The config file
  -t, --target <TARGET>        The target to process
  -m, --mapping <MAPPING>      The mapping file
  -a, --api-proxy <API_PROXY>  The user file
  -s, --server                 Run in server mode
  -l, --log-level <LOG_LEVEL>  log level
  -h, --help                   Print help
  -V, --version                Print version
```

**Reply by euzu** [reply, marked as answer]
it is really easy if you know how

```sh
rust install cross
env  RUSTFLAGS="--remap-path-prefix $HOME=~" cross build --release --target armv7-unknown-linux-musleabihf
```

**Comment by euzu** [comment]
you need to create your config files,  and copy the web folder.
